Add a simple json schema for adding services to dittojay through json

brendanlensink opened the pull request:
### Fixes: #225

### Summary of Problem:
We'd like to be able to add and manage services created with Dittojay through JSON instead of having to add them in code.

### Proposed Solution:
Created a simple JSON schema that can be parsed to create services from.

A couple questions:
- I noticed that the `CBMutableService` constructor calls the UUID param `type`, is there a reason we should do this as well, or is it clearer to just stick with `uuid`?
- Is it helpful to have some sort of user-defined name attribute for keeping track of services once they're loaded?
- Do we need to require UUIDs be set ahead of the time, or can they be optional and generated at runtime by Dittojay?
- Does it make sense to have `value` passed in as a String that we can then convert into `Data?`?

brendanlensink committed:
add a simple json schema for adding services to dittojay through json

sakuraehikaru reviewed (changes requested):
Thanks Brendan!

1. I don't think we should merge this to master. I am open to suggestions, but my initial thought is to create a new feature branch with the version number bumped as well.
2. I am not sure why it's called type, but it may have to do with how CBUUID is a more specialized version of plain UUID, where it's able to recognize shorter strings of UUID that are specific to Bluetooth, and can transform those shorter UUIDs into their full length strings under the hood.
3. I think we can allow specifying name and even description in the schema, because while CoreBluetooth doesn't keep track of those two properties, Dittojay will have to keep track of the added services and characteristics anyway, so I think we can add whatever metadata we want. Extra metadata also just seems useful in many cases: debugging, rendering nicer UI, etc...
4. I think it is safe to make UUID optional. Though extremely unlikely, when we generate the UUIDs we just want to avoid accidentally generating a commonly-used UUID predefined by The Bluetooth Special Interest Group (SIG). Based on this [documentation](https://developer.apple.com/documentation/corebluetooth/cbuuid), I think we can minimize the chance of that by always generating a full 128 bit UUID, since common Bluetooth UUIDs seem to be 16 or 32 bit short form.
5. As for defining a characteristic's value, I'm gonna loop in @nbrooke here because I might not be able to picture the full implementation - but here are some of my thoughts:
- Instead of just `value:<a big chunk of base64strings>`, we might want to allow people to define the mapping of the data that will correspond to the Swift struct they will implement. For example, maybe `value:<another json object>` could help map the following data structure better, and would definitely make it easier to modify the schema later.
```
struct NumOfLog: Receivable {
    var startByte =         UInt8(0)
    var payloadLength =     UInt8(0)
    var responseCode =      UInt8(0)
    var data =              UInt16(0)
    var crc =               UInt16(0)
}
```
- I am also wondering if we should come up with a trick, if possible, to separate the base64 strings from the main schema. The schema can become really hard to read if it's littered with unreadable strings. Perhaps the actual value in the schema will just be a key/link to a different file, whose sole purpose is to just contain a list of binary data to be used by the main schema file.
- We may also want to allow defining the endianess of the data, though that field can be optional and use a sensible default
  - The endianess may also need to be on a per-field basis, just because I've encountered devices with data that has mixed endianess 😂

brendanlensink committed:
add description metadata to service and characteristic fields
change value to contain a JSON structure to make viewing the schema clearer

brendanlensink commented:
> I don't think we should merge this to master.

That's totally reasonable, would this end up being version 0.7.2 then?

> I think we can allow specifying name and even description

Description is a great idea, added that as well.

> Instead of just `value` ...

Changed `value` to hold another JSON object, with the intent being that Dittojay will parse the JSON and concat all the values to create the Data to be read.

I think it would be neat to create some sort of framework to separate the `value` data and the schema, but it might be a more 2.0/next iteration feature if that makes sense?

JeremyChiang commented:
Yes, let's defer any high complexity features that might not be necessary to have right away :)

nbrooke commented:
`value` being an object might not work (or at least, not work by itself). 

The fields are going to be order dependant (except in the trivial case of only one field), but I don't think that ordering is well defined for JSON (like once we have loaded that in, there is no way to say "give me the fields in order", which is pretty essential for creating the binary data to send).

What I think we want is an array that defines things like ordering and byte width:

```
"value" : [
   { "byteWidth": 2, "value": 700 },
   { "byteWidth": 4, "value": -100 }
]
```

This is (I believe) the simplest thing that could work. We can get more complicated by adding possible options to the fields (`{"byteWidth": 2, "bigEndian": true, "value": 700 }`) or by naming the fields so we can do the more seperate-schema-and-data approach later:

```
"schema" : [
   { "name" : "foo", "byteWidth": 2, "bigEndian": true}.
   { "name" : "bar", "byteWidth": 4, "bigEndian": false}
],
data: {
  "foo": 700,
  "bar": -100
}
```

brendanlensink commented:
@sakuraehikaru Is it worth keeping this PR open, or can we close this and re-open if we want to address this in the future?

sakuraehikaru commented:
@brendanlensink I have no preference, other than if we do close this, then let's **not** delete this branch so we don't lose any work that's already done.
